src/image.ts

```diff
@@ -125,27 +131,66 @@ function decodeQrBuffer(fileInfo: FileInfo, log: Log): string | undefined {
-        }
 
-        // let's make sure the QR code's version is tight
-        try {
-            const qrCode = create(code.data, { errorCorrectionLevel: 'low' });
-            if (qrCode.version < code.version) {
-                log.warn(`QR code has version ${code.version}, but could have been created with version ${qrCode.version} (with low error-correcting level). Make sure the larger version was chosen on purpose (e.g., not hardcoded).`, ErrorCode.INVALID_QR_VERSION);
+    // decide whether is a shc or shlink or unknown
+
+    if (!code.chunks) {
+        log.fatal(`No segments in QR code. Expected 1-2`, ErrorCode.INVALID_QR);
+        return undefined;
+    } 
+
+    // if this is an 'shc' check for 2 correct segments
+    // this check will fail on a really poorly formed shc
+    if (isShc(code.data)) {
+        if (code.chunks.length !== 2) {
+            log.error(
+                `Wrong number of segments in QR code: found ${code.chunks.length}, expected 2` +
+                    `\nSegments types: ${code.chunks.map((chunk, i) => `${i + 1}: ${chunk.type}`).join("; ")}`,
+                ErrorCode.INVALID_QR
+            );
+        } else {
+            if (code.chunks[0].type !== "byte") {
+                // unlikely, since 'shc:/' can only be legally encoded as with byte mode;
+                // was not able to create test case for this
+                log.error(
+                    `Wrong encoding mode for first QR segment: found ${code.chunks[0].type}, expected "byte"`,
+                    ErrorCode.INVALID_QR
+                );
+            }
+            if (code.chunks[1].type !== "numeric") {
+                log.error(
+                    `Wrong encoding mode for second QR segment: found ${code.chunks[0].type}, expected "numeric"`,
+                    ErrorCode.INVALID_QR
+                );
+            }
+
+            // let's make sure the QR code's version is tight
+            try {
+                const qrCode = create(code.data, { errorCorrectionLevel: "low" });
+                if (qrCode.version < code.version) {
+                    log.warn(
+                        `QR code has version ${code.version}, but could have been created with version ${qrCode.version} (with low error-correcting level). Make sure the larger version was chosen on purpose (e.g., not hardcoded).`,
+                        ErrorCode.INVALID_QR_VERSION
+                    );
+                }
+            } catch (err) {
+                log.warn(`Can't re-create QR to check optimal version choice: ${(err as Error).message})`);
             }
-        } catch (err) {
-            log.warn(`Can't re-create QR to check optimal version choice: ${(err as Error).message})`);
+        }
+    } else /* not a 'shc' -- assuming 'shlink' */ {
+        if (code.chunks.length !== 1) {
+            log.error(
```

Comment:
An error here may be a bit harsh; we don't specify anything about the number of chunks, right? Maybe warn?

Reply:
We can do a warning here. 

Before SHL, there was just one kind of expected input and we could just push the decoded data further into the validator to tell you why it was wrong.
Now we have to determine what kind of thing we have first, SHC or SHL before we continue.  So, for example, if the header is wrong 'shj:' then we don't know which type to validate - so I was making a decision based on the number of chunks. If it wasn't 1, then you must be trying to validate something other than a SML.  
And I did not want to process every malformed SHC as an SHL by default.

I'll revisit the logic here.